File: Poisson_test/create_RHS.m
function [Rhs,Nx,Nz]=create_RHS(Displ)

% This function creates the Rhs of the Poisson's equation. 
% The Rhs is obtained summing dx+dy (defined as Displ.u and Displ.v).

u=Displ.u;
w=Displ.v;
x=Displ.x;
z=Displ.y;

[width,height]=size(u);

% Preallocation for the variables du and dw;
du=zeros(width-2,height-2);
dw=zeros(width-2,height-2);

% Compute the central differences 
for k=2:width-1
for j=2:height-1
    
    du(k,j)=(u(k+1,j)-u(k-1,j))/2*(abs(x(2,2)-x(2,1)));
    dw(k,j)=(w(k,j+1)-w(k,j-1))/2*(abs(z(1,1)-z(1,2)));

end
end

Rhs=(du+dw);
Rhs(isnan(Rhs)) = 0 ;

[Nz,Nx]=size(Rhs);

% create_RHS can be replaced using simply the function gradient()... 

File: create_rhs.py
import numpy as np

def create_RHS(Displ):
    """
    Create the right-hand side of the Poisson equation
    
    Parameters:
    -----------
    Displ : dict
        Dictionary containing displacement field data (x, y, u, v)
        
    Returns:
    --------
    Rhs : ndarray
        Right-hand side of the Poisson equation
    Nx : int
        Width of the RHS field
    Nz : int
        Height of the RHS field
        
    Notes:
    ------
    This function computes the central difference fields for the displacement
    components and combines them to form the RHS of the Poisson equation.
    """
    # Extract displacement components
    u = Displ['u']
    w = Displ['v']
    x = Displ['x']
    z = Displ['y']
    
    # Get dimensions
    width, height = u.shape
    
    # Preallocate arrays for derivatives
    du = np.zeros((width-2, height-2))
    dw = np.zeros((width-2, height-2))
    
    # Compute central differences
    for k in range(1, width-1):
        for j in range(1, height-1):
            du[k-1, j-1] = (u[k+1, j] - u[k-1, j]) / 2 * abs(x[1, 1] - x[1, 0])
            dw[k-1, j-1] = (w[k, j+1] - w[k, j-1]) / 2 * abs(z[0, 0] - z[0, 1])
    
    # Combine derivatives to form RHS
    Rhs = du + dw
    
    # Replace NaN values with zeros
    Rhs[np.isnan(Rhs)] = 0
    
    # Get dimensions of RHS
    Nz, Nx = Rhs.shape
    
    return Rhs, Nx, Nz
